pkg/netscan/discover.go

```diff
@@ -0,0 +1,271 @@
+			// or more ports are scanned.
+			probeFactor*params.Timeout*time.Duration(math.Min(float64(portCount), 3)))
+	}
+	params.Logger.Debugf("total estimated network probes: %d, async limit: %d, probe timeout: %v, estimated time: %s",
+		estimatedProbes, asyncLimit, params.Timeout, estimatedTimeStr)
+
+	ipCh := make(chan uint32, asyncLimit)
+	resultCh := make(chan []ProbeResult)
+
+	wParams := workerParams{
+		Params:   params,
+		ipCh:     ipCh,
+		resultCh: resultCh,
+		ctx:      ctx,
+		proto:    proto,
+	}
+
+	// start the workers before adding any ips, so they are ready to process
+	var wgIPWorkers sync.WaitGroup
+	wgIPWorkers.Add(asyncLimit)
+	for i := 0; i < asyncLimit; i++ {
+		go func() {
+			defer wgIPWorkers.Done()
+			ipWorker(wParams)
+		}()
+	}
+
+	go func() {
+		var wgIPGenerators sync.WaitGroup
+		for _, ipnet := range ipnets {
+			select {
+			case <-ctx.Done():
+				// quit early if we have been cancelled
+				return
+			default:
+			}
+
+			// wait on each ipGenerator
+			wgIPGenerators.Add(1)
+			go func(inet *net.IPNet) {
+				defer wgIPGenerators.Done()
+				ipGenerator(ctx, inet, ipCh)
+			}(ipnet)
+		}
+
+		// wait for all ip generators to finish, then we can close the ip channel
+		wgIPGenerators.Wait()
+		close(ipCh)
+
+		// wait for the ipWorkers to finish, then close the results channel which
+		// will let the enclosing function finish
+		wgIPWorkers.Wait()
+		close(resultCh)
+	}()
+
+	// this blocks until the resultCh is closed in above go routine
+	return processResultChannel(resultCh, proto, params)
+}
+
+// processResultChannel reads all incoming results until the resultCh is closed.
+// it determines if a device is new or existing, and proceeds accordingly.
+//
+// Does not check for context cancellation because we still want to
+// process any in-flight results.
+func processResultChannel(resultCh chan []ProbeResult, proto ProtocolSpecificDiscovery, params Params) []DiscoveredDevice {
+	devices := make([]DiscoveredDevice, 0)
+	for probeResults := range resultCh {
+		if len(probeResults) == 0 {
+			continue
+		}
+
+		for _, probeResult := range probeResults {
+			dev, err := proto.ConvertProbeResult(probeResult, params)
+			if err != nil {
+				params.Logger.Warnf("issue converting probe result to discovered device: %s", err.Error())
+				continue
+			}
+			// only add if a valid device was returned
+			if dev.Name != "" {
+				devices = append(devices, dev)
+			}
+		}
+	}
+	return devices
+}
+
+func handleConnectionInternal(host string, port string, conn net.Conn, params workerParams) {
+	// on udp, the dial is always successful, so don't print
+	if !strings.HasPrefix(params.NetworkProtocol, NetworkUDP) {
+		params.Logger.Debugf("Connection dialed %s://%s:%s", params.NetworkProtocol, host, port)
+	}
+
+	results, err := params.proto.OnConnectionDialed(host, port, conn, params.Params)
+	if err != nil {
+		params.Logger.Debugf(err.Error())
```

Comment:
Comment why Debug and Not Error or Warn log message

```suggestion
		params.Logger.Debug(err.Error())
```

Reply:
Will comment about why debug.